### PR TITLE
Make handling of resolution of remoteFilter more robust under various timings

### LINKED ISSUE #300: Invoking callback passed to remoteFilter in various timings breaks At.js
___The code provided here are meant to be ran in a page where At.js is already loaded, like one of the pages from the examples directory___

##### If an older callback passed to `remoteFilter()` is invoked after a newer one, the suggestion list will disappear (textarea and contentEditable)

1. Run the following in console

    ```js
    (function ($) {
        var textArea = document.createElement('TEXTAREA');
        document.body.appendChild(textArea);
        $(textArea).atwho({
            at: "@",
            data:[],
            displayTimeout: 0,
            callbacks: {
                remoteFilter: function (query, cb) {
                    var timeout = query === 'gon' ? 1000 : 0;
                    window.setTimeout(function () {
                        cb([query]);
                    }, timeout);
                }
            }
        });
    })(jQuery);
    ```
2. scroll down to the bottom of the page and type `@gone` into the textarea
3. the suggestion list will show `@gone` for a bit, then disappear

Typing query that doesn't start with gon works as expected. (Note: if you type `@gone` super quickly, the bug won't trigger)

##### if the callback is invoked after the key char has been deleted, At.js throws `TypeError` for textarea and renders the list on top left of the viewport for contentEditable

1. Run the following in console

    ```js
    (function ($) {
        var textArea = document.createElement('TEXTAREA');
        document.body.appendChild(textArea);
        $(textArea).atwho({
            at: "@",
            data:[],
            displayTimeout: 0,
            callbacks: {
                remoteFilter: function (query, cb) {
                    window.setTimeout(function () {
                        cb([query]);
                    }, 1000);
                }
            }
        });
    })(jQuery);
    ```
2. type `@` then delete it immidiately
3. after 1 second, look in the console

Same steps for contentEditable, the code for it is below. The div is invisible when not clicked, it should be at the bottom left corner.

```js
(function ($) {
    var editable = document.createElement('DIV');
    editable.contentEditable = 'true';
    editable.style.width = '100px';
    editable.style.height = '100px';
    document.body.appendChild(editable);
    $(editable).atwho({
        at: "@",
        data:[],
        displayTimeout: 0,
        callbacks: {
            remoteFilter: function (query, cb) {
                window.setTimeout(function () {
                    cb([query]);
                }, 1000);
            }
        }
    });
})(jQuery);
```

I suspect this is what caused #235

Thank you for you hard work! 辛苦了!

### PR DESCRIPTION
**Pull request by XrXr**
See #300 
Closes #300

### CODE REVIEW

**Comment by ichord** [comment]
First of all, thanks!!

I just see `expectedQueryCBId` is assign with `{}` and `null`; what else should be?
And please run grunt to produce `*.js` file, so i can merge it.

**Comment by XrXr** [comment]
`{}` is being used as the id since different objects will compare false with `===`. It doesn't really matter what is inside the object, it's just a new object so it compare false with any previous values.

When `expectedQueryCBId` is null, any `{}` won't be accepted. It's a mean for ignoring all resolution from `remoteFilter`

So expectedQueryCBId should only ever be an object or null

**Comment by ichord** [comment]
Why  `expectedQueryCBId` is not just a boolean? What is the benefit to be a object?
At least, I don't quit understand If you don't explain it.

**Comment by XrXr** [comment]
Each call to lookUp initiate an async task through `remoteFilter()`, these task can resolve in different order.

Suppose we have calls to `_lookup()` `a` `b` and `c`, in this order. The resolution order, or the the order that each `remoteFilter()` invoke the callback passed to it can be:

`a b c`
`b c a`
`c b a`
`c a b`
...

We only want to response to the resolution of `c`, but there is no guarantee that it comes back last.
It shouldn't be hard to see why a simple true and false is not enough for this, since if it represent whether we process the callback or not, `a` might come back first, and the result from `c` gets lost, which is not what we want.

in https://github.com/ichord/At.js/pull/301/files#diff-16defa3b7d5b7c6991284c260731a1ccR137
```coffee
@expectedQueryCBId = {}
@model.query query.text, $.proxy(_callback, this, @expectedQueryCBId)
```

`@expectedQueryCBId = {}` this is creating a new object every time. This is very different from
`@expectedQueryCBId = null` or `@expectedQueryCBId = false`. `null` and `false` are singletons,
if you do
```js
var a = false;
var b = false;
a === b  // true
```
but if you
```js
var a = {}
var b = {}
a === b // false
a === a // true
```

I'm leveraging the uniqueness of each object as a mean for identifying each remoteFilter initiation.

The next line  `@model.query query.text, $.proxy(_callback, this, @expectedQueryCBId)` is scheduling `_callback` to be called with the current `expectedQueryCBId`. The call to `$.proxy` is evaluated before the call to `@model.query`. By the time `_callback()` is actually invoked `@expectedQueryCBId` might have been changed already, but the first argument it receives is still the object we created back when it is the most recent id.

The code would be more readable if I id each callback with a number and keep incrementing it, however when the number gets too large it becomes `Infinity`. It's not really a realistic probability since that number is `1.7976931348623157E+10308` but I was having a bit of fun.

tl;dr sorry, I did a little bit of cleaver coding, I can change the id to be a number if you'd like. The code would be less terse but more readable.

**Comment by XrXr** [comment]
I also found a case with remoteFilter resolving after input has blured. I've added the fix to the patch.

**Comment by ichord** [comment]
Cool! understood!  I think we could take the middle way of it: 
Add a function named `_generateID` or another better name to return an "ID" which we will return a new empty object `{}` here. 
So that we could change the returns value later and make it more readable.
How about it?

**Comment by XrXr** [comment]
Good call! I made the change.

**Comment by seriousben** [comment]
Can we get a release?

**Comment by ichord** [comment]
@seriousben released.
